### DISCUSSION TITLE
Trying to hack up a change to fdpp kernel, but getting a crash.

### DISCUSSION THREAD

**Discussion by andrewbird** [dosemu2/fdpp#248] [Q&A] [answered]
Firstly I'm still without broadband here, so I'm having to retype everything on mobile. I have a structure that I'd like to be able to set es:Di to. Simplified the code looks something like this:

Struct xgetxxx s;
Struct xgetxxx FAR *p;

p=MK_FAR(s);

regs.es = FP_SEG(p);
regs.di = FP_OFF(p);

But I get a crash, fdpp logging in dosemu mentions something about rtti symbol not found for class. What am I doing wrong?

Thanks.

**Comment by stsp** [comment, marked as answer]
Never seen any rtti-mentioning errors,
so no idea really. This may be a bad build.
But make sure `regs` has the same
life-range as `s` and use MK_xxx_OBJ()
macros:
https://github.com/dosemu2/fdpp/blob/master/kernel/initdisk.c#L728-L733

**Reply by andrewbird** [reply]
Unfortunately switching to the obj macros didnt help. I may have messed up the parameter passing to the function, perhaps this is corrupting something. But thanks anyway!

**Reply by andrewbird** [reply]
Actually using the MK_xxx_OBJ() did fix this problem, only I had something else wrong afterwards that also caused a crash and I didn't notice the difference.

**Comment by andrewbird** [comment]
Can you remind me what I need to do to debug FDPP with gdb, please? I know you've told me in the past, but I've just tried searching old PRs and issues and can't find it. That's a lesson learnt for me, and so this time I'll add it to dosemu's wiki!

**Reply by stsp** [reply]
What exactly aspect of debugging?
You can just run `dosemu -gdb` and
set break-points to fdpp functions.
Many fdpp symbols require double
underscore to find their DOS address,
after which you can use `mem_base`
to see the actual content.

**Reply by andrewbird** [reply]
I had tried that but got bogged down before hitting my breakpoint in this:
~~~
Thread 1 "dosemu.bin" received signal SIG34, Real-time event 34.
0x00007ffff7d18bcf in __GI___poll (fds=0x7fff219676f8, nfds=1, timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
29	../sysdeps/unix/sysv/linux/poll.c: No such file or directory.
(gdb) cont
Continuing.

Thread 1 "dosemu.bin" received signal SIG34, Real-time event 34.
futex_wait (private=0, expected=2, futex_word=0x5555568c66c0 <pic_mtx>) at ../sysdeps/nptl/futex-internal.h:146
146	../sysdeps/nptl/futex-internal.h: No such file or directory.
(gdb) 
Continuing.

Thread 1 "dosemu.bin" received signal SIG34, Real-time event 34.
futex_wait (private=0, expected=2, futex_word=0x7ffff7e4bf58 <rend_mtx>) at ../sysdeps/nptl/futex-internal.h:146
146	in ../sysdeps/nptl/futex-internal.h
(gdb) 
Continuing.

Thread 1 "dosemu.bin" received signal SIG34, Real-time event 34.
futex_wait (private=0, expected=2, futex_word=0x5555568c66c0 <pic_mtx>) at ../sysdeps/nptl/futex-internal.h:146
146	in ../sysdeps/nptl/futex-internal.h
(gdb) 
Continuing.
~~~

I kept doing 'continue' but it felt like I wasn't getting anywhere.

Also I remember setting EXTRA_DEBUG=1 before but I think you mentioned that it doesn't work with the meson build system?

**Reply by stsp** [reply]
`handle SIG34 nostop noprint`
You can also put that into .gdbinit
Will look into EXTRA_DEBUG for meson.

**Reply by andrewbird** [reply]
Yes I almost had that, except I had SIGSEGV instead of SIG34. I'm now able to 'land' exactly on my new function, thanks!

**Reply by stsp** [reply]
I added EXTRA_DEBUG support.
It is enabled by default, but you
can do
`./configure.meson --buildtype=release build`

**Reply by andrewbird** [reply]
Thanks, I now see your commit fae984d257. I see it looks for string 'debug' in the buildtype, so I presume I use 'debug' instead of 'release' in the line above?

**Reply by stsp** [reply]
No, as it seems to be a default anyway.
Only `release` changes settings (in a way
you don't need).

**Reply by andrewbird** [reply]
So I should omit the `--buildtype=release` altogether?

**Reply by stsp** [reply]
Seems so.

**Reply by andrewbird** [reply]
Cool, thanks!
